File: common/src/main/java/net/polyv/android/player/common/ui/component/PLVMediaPlayerLongPressSpeedHintLayout.java
package net.polyv.android.player.common.ui.component;

import static net.polyv.android.player.sdk.foundation.lang.PreconditionsKt.requireNotNull;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.TextView;

import net.polyv.android.player.common.R;
import net.polyv.android.player.common.di.PLVMediaPlayerLocalProvider;
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.PLVMPMediaControllerViewModel;
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.viewstate.PLVMPMediaControllerViewState;
import net.polyv.android.player.sdk.foundation.lang.PLVRememberState;
import net.polyv.android.player.sdk.foundation.lang.PLVRememberStateCompareResult;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;

/**
 * @author Hoshiiro
 */
public class PLVMediaPlayerLongPressSpeedHintLayout extends FrameLayout {

    private TextView speedTv;
    private TextView speedHintTv;

    private boolean isVisible = false;

    public PLVMediaPlayerLongPressSpeedHintLayout(@NonNull Context context) {
        super(context);
    }

    public PLVMediaPlayerLongPressSpeedHintLayout(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public PLVMediaPlayerLongPressSpeedHintLayout(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    {
        LayoutInflater.from(getContext()).inflate(R.layout.plv_media_player_ui_component_long_press_speed_hint_layout, this);
        speedTv = findViewById(R.id.plv_media_player_speed_tv);
        speedHintTv = findViewById(R.id.plv_media_player_speed_hint_tv);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        if (isInEditMode()) return;

        requireNotNull(PLVMediaPlayerLocalProvider.localDependScope.on(this).current())
                .get(PLVMPMediaControllerViewModel.class)
                .getMediaControllerViewState()
                .observeUntilViewDetached(this, new Function1<PLVMPMediaControllerViewState, Unit>() {
                    @Override
                    public Unit invoke(PLVMPMediaControllerViewState viewState) {
                        isVisible = viewState.getLongPressSpeeding();
                        onViewStateChanged();
                        return null;
                    }
                });
    }

    protected void onViewStateChanged() {
        PLVRememberState.rememberStateOf(this, "updateLongPressSpeedHint")
                .compareLastAndSet(isVisible)
                .ifNotEquals(new Function1<PLVRememberStateCompareResult, Unit>() {
                    @Override
                    public Unit invoke(PLVRememberStateCompareResult result) {
                        updateLongPressSpeedHint();
                        return null;
                    }
                });
    }

    protected void updateLongPressSpeedHint() {
        setVisibility(isVisible ? View.VISIBLE : View.GONE);
    }

}

File: common/src/main/java/net/polyv/android/player/common/ui/component/gesture/PLVMediaPlayerLongPressSpeedHintLayout.kt
package net.polyv.android.player.common.ui.component.gesture

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.FrameLayout
import android.widget.TextView
import net.polyv.android.player.common.R
import net.polyv.android.player.common.di.PLVMediaPlayerLocalProvider
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.PLVMPMediaControllerViewModel
import net.polyv.android.player.common.modules.mediacontroller.viewmodel.viewstate.PLVMPMediaControllerViewState

/**
 * @author Hoshiiro
 */
class PLVMediaPlayerLongPressSpeedHintLayout @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : FrameLayout(context, attrs, defStyleAttr) {

    private val speedTv by lazy { findViewById<TextView>(R.id.plv_media_player_speed_tv) }
    private val speedHintTv by lazy { findViewById<TextView>(R.id.plv_media_player_speed_hint_tv) }

    init {
        LayoutInflater.from(context).inflate(R.layout.plv_media_player_ui_component_long_press_speed_hint_layout, this)
    }

    override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        if (isInEditMode) return

        val dependScope = PLVMediaPlayerLocalProvider.localDependScope.on(this).current()!!
        dependScope.get<PLVMPMediaControllerViewModel>()
            .mediaControllerViewState
            .observe { viewState: PLVMPMediaControllerViewState ->
                visibility = if (viewState.longPressSpeeding) VISIBLE else GONE
            }
            .disposeOnDetached(this)
    }

}
